In [146]:
import pandas as pd
train = pd.read_csv('./dataset/train.csv')
test = pd.read_csv('./dataset/test.csv')
sample_submission = pd.read_csv('./dataset/sample_submission.csv')
train.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6255 entries, 0 to 6254
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            6255 non-null   int64  
 1   Store         6255 non-null   int64  
 2   Date          6255 non-null   object 
 3   Temperature   6255 non-null   float64
 4   Fuel_Price    6255 non-null   float64
 5   Promotion1    2102 non-null   float64
 6   Promotion2    1592 non-null   float64
 7   Promotion3    1885 non-null   float64
 8   Promotion4    1819 non-null   float64
 9   Promotion5    2115 non-null   float64
 10  Unemployment  6255 non-null   float64
 11  IsHoliday     6255 non-null   bool   
 12  Weekly_Sales  6255 non-null   float64
dtypes: bool(1), float64(9), int64(2), object(1)
memory usage: 592.6+ KB


In [147]:
train.head()

id  Store        Date  Temperature  Fuel_Price  Promotion1  Promotion2  \
0   1      1  05/02/2010        42.31       2.572         NaN         NaN   
1   2      1  12/02/2010        38.51       2.548         NaN         NaN   
2   3      1  19/02/2010        39.93       2.514         NaN         NaN   
3   4      1  26/02/2010        46.63       2.561         NaN         NaN   
4   5      1  05/03/2010        46.50       2.625         NaN         NaN   

   Promotion3  Promotion4  Promotion5  Unemployment  IsHoliday  Weekly_Sales  
0         NaN         NaN         NaN         8.106      False    1643690.90  
1         NaN         NaN         NaN         8.106       True    1641957.44  
2         NaN         NaN         NaN         8.106      False    1611968.17  
3         NaN         NaN         NaN         8.106      False    1409727.59  
4         NaN         NaN         NaN         8.106      False    1554806.68

In [148]:
train.tail()

id  Store        Date  Temperature  Fuel_Price  Promotion1  \
6250  6251     45  31/08/2012        75.09       3.867    23641.30   
6251  6252     45  07/09/2012        75.70       3.911    11024.45   
6252  6253     45  14/09/2012        67.87       3.948    11407.95   
6253  6254     45  21/09/2012        65.32       4.038     8452.20   
6254  6255     45  28/09/2012        64.88       3.997     4556.61   

      Promotion2  Promotion3  Promotion4  Promotion5  Unemployment  IsHoliday  \
6250        6.00       92.93     6988.31     3992.13         8.684      False   
6251       12.80       52.63     1854.77     2055.70         8.684       True   
6252         NaN        4.30     3421.72     5268.92         8.684      False   
6253       92.28       63.24     2376.38     8670.40         8.684      False   
6254       20.64        1.50     1601.01     3288.25         8.684      False   

      Weekly_Sales  
6250     734297.87  
6251     766512.66  
6252     702238.27  
6253     723086.20  
6254     713173.95

6255개의 인덱스, 13개의 컬럼, 45개의 store, 2010-02-05 ~ 2012-09-28 date
Promotion 부분에 결측치 확인
date, IsHoliday 의 타입이 object, boolean 이기 때문에 데이터 전처리 필요

In [149]:
train = train.fillna(0)
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6255 entries, 0 to 6254
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            6255 non-null   int64  
 1   Store         6255 non-null   int64  
 2   Date          6255 non-null   object 
 3   Temperature   6255 non-null   float64
 4   Fuel_Price    6255 non-null   float64
 5   Promotion1    6255 non-null   float64
 6   Promotion2    6255 non-null   float64
 7   Promotion3    6255 non-null   float64
 8   Promotion4    6255 non-null   float64
 9   Promotion5    6255 non-null   float64
 10  Unemployment  6255 non-null   float64
 11  IsHoliday     6255 non-null   bool   
 12  Weekly_Sales  6255 non-null   float64
dtypes: bool(1), float64(9), int64(2), object(1)
memory usage: 592.6+ KB


fillna(0) 메소드를 사용하여 결측치인 부분 0으로 채우기

In [150]:
def get_date(date):
    newDate = date[6:10]
    newDate += date[3:5]
    newDate += date[0:2]
    return int(newDate)
train['Date'] =  train['Date'].apply(get_date)
train

id  Store      Date  Temperature  Fuel_Price  Promotion1  Promotion2  \
0        1      1  20100205        42.31       2.572        0.00        0.00   
1        2      1  20100212        38.51       2.548        0.00        0.00   
2        3      1  20100219        39.93       2.514        0.00        0.00   
3        4      1  20100226        46.63       2.561        0.00        0.00   
4        5      1  20100305        46.50       2.625        0.00        0.00   
...    ...    ...       ...          ...         ...         ...         ...   
6250  6251     45  20120831        75.09       3.867    23641.30        6.00   
6251  6252     45  20120907        75.70       3.911    11024.45       12.80   
6252  6253     45  20120914        67.87       3.948    11407.95        0.00   
6253  6254     45  20120921        65.32       4.038     8452.20       92.28   
6254  6255     45  20120928        64.88       3.997     4556.61       20.64   

      Promotion3  Promotion4  Promotion5  Unemployment  IsHoliday  \
0           0.00        0.00        0.00         8.106      False   
1           0.00        0.00        0.00         8.106       True   
2           0.00        0.00        0.00         8.106      False   
3           0.00        0.00        0.00         8.106      False   
4           0.00        0.00        0.00         8.106      False   
...          ...         ...         ...           ...        ...   
6250       92.93     6988.31     3992.13         8.684      False   
6251       52.63     1854.77     2055.70         8.684       True   
6252        4.30     3421.72     5268.92         8.684      False   
6253       63.24     2376.38     8670.40         8.684      False   
6254        1.50     1601.01     3288.25         8.684      False   

      Weekly_Sales  
0       1643690.90  
1       1641957.44  
2       1611968.17  
3       1409727.59  
4       1554806.68  
...            ...  
6250     734297.87  
6251     766512.66  
6252     702238.27  
6253     723086.20  
6254     713173.95  

[6255 rows x 13 columns]

Date col 을 추출해서 YYYYMMDD 의 숫자형태로 변환

In [151]:
def holiday_to_number(isholiday):
    number = 1 if isholiday == True else 0
    return number
train['IsHoliday'] = train['IsHoliday'].apply(holiday_to_number)
train

id  Store      Date  Temperature  Fuel_Price  Promotion1  Promotion2  \
0        1      1  20100205        42.31       2.572        0.00        0.00   
1        2      1  20100212        38.51       2.548        0.00        0.00   
2        3      1  20100219        39.93       2.514        0.00        0.00   
3        4      1  20100226        46.63       2.561        0.00        0.00   
4        5      1  20100305        46.50       2.625        0.00        0.00   
...    ...    ...       ...          ...         ...         ...         ...   
6250  6251     45  20120831        75.09       3.867    23641.30        6.00   
6251  6252     45  20120907        75.70       3.911    11024.45       12.80   
6252  6253     45  20120914        67.87       3.948    11407.95        0.00   
6253  6254     45  20120921        65.32       4.038     8452.20       92.28   
6254  6255     45  20120928        64.88       3.997     4556.61       20.64   

      Promotion3  Promotion4  Promotion5  Unemployment  IsHoliday  \
0           0.00        0.00        0.00         8.106          0   
1           0.00        0.00        0.00         8.106          1   
2           0.00        0.00        0.00         8.106          0   
3           0.00        0.00        0.00         8.106          0   
4           0.00        0.00        0.00         8.106          0   
...          ...         ...         ...           ...        ...   
6250       92.93     6988.31     3992.13         8.684          0   
6251       52.63     1854.77     2055.70         8.684          1   
6252        4.30     3421.72     5268.92         8.684          0   
6253       63.24     2376.38     8670.40         8.684          0   
6254        1.50     1601.01     3288.25         8.684          0   

      Weekly_Sales  
0       1643690.90  
1       1641957.44  
2       1611968.17  
3       1409727.59  
4       1554806.68  
...            ...  
6250     734297.87  
6251     766512.66  
6252     702238.27  
6253     723086.20  
6254     713173.95  

[6255 rows x 13 columns]

Boolean 으로 되어있는 IsHoliday 를 0,1 의 숫자형태로 변환

In [152]:
test = test.fillna(0)
test['Date'] = test['Date'].apply(get_date)
test['IsHoliday'] = test['IsHoliday'].apply(holiday_to_number)
test

id  Store      Date  Temperature  Fuel_Price  Promotion1  Promotion2  \
0      1      1  20121005        68.55       3.617     8077.89        0.00   
1      2      1  20121012        62.99       3.601     2086.18        0.00   
2      3      1  20121019        67.97       3.594      950.33        0.00   
3      4      1  20121026        69.16       3.506     2585.85       31.75   
4      5      2  20121005        70.27       3.617     6037.76        0.00   
..   ...    ...       ...          ...         ...         ...         ...   
175  176     44  20121026        46.97       3.755       38.65        2.61   
176  177     45  20121005        64.89       3.985     5046.74        0.00   
177  178     45  20121012        54.47       4.000     1956.28        0.00   
178  179     45  20121019        56.47       3.969     2004.02        0.00   
179  180     45  20121026        58.85       3.882     4018.91       58.08   

     Promotion3  Promotion4  Promotion5  Unemployment  IsHoliday  
0         18.22     3617.43     3626.14         6.573          0  
1          8.11      602.36     5926.45         6.573          0  
2          4.93       80.25     2312.85         6.573          0  
3          6.00     1057.16     1305.01         6.573          0  
4         10.04     3027.37     3853.40         6.170          0  
..          ...         ...         ...           ...        ...  
175        0.98        0.00      457.74         5.217          0  
176       18.82     2253.43     2340.01         8.667          0  
177        7.89      599.32     3990.54         8.667          0  
178        3.18      437.73     1537.49         8.667          0  
179      100.00      211.94      858.33         8.667          0  

[180 rows x 12 columns]

test set 도 data processing

In [153]:
from sklearn.linear_model import LinearRegression

# 모델 선언
model = LinearRegression()

# 분석 의미 없는 id 제거
train = train.drop(columns=['id'])
test = test.drop(columns=['id'])

# 학습에 사용할 정보와 예측하고자 하는 정보를 분리
x_train = train.drop(columns=['Weekly_Sales'])
y_train = train['Weekly_Sales']

model.fit(x_train,y_train)

LinearRegression()

fit()메소드를 사용해 모델 학습

In [154]:
prediction = model.predict(test)
print(prediction)

[1285506.9104259  1232370.26168644 1122499.82146025 1101443.28590918
 1244055.18729818 1331340.5675348  1188846.31519544 1212635.38487136
 1084495.50105548 1055718.61522686 1107928.82202089 1013007.67883384
 1337047.39753509 1312551.03650224 1325568.23152494 1256841.9999305
 1106653.23493409 1131280.44433832 1055060.95868158 1035959.09844303
 1324725.76512444 1245668.83768654 1129031.63730729 1160664.47498488
 1109121.18632925 1051002.65329385 1162851.52769041 1092438.84732485
 1121451.86383891 1130048.61292243 1097720.09002829 1057879.84470081
 1070219.79638255 1065013.06833756 1057719.56199849  997793.0376519
 1350490.02370596 1390337.88095915 1280650.10906935 1331612.80350721
 1155709.16739166 1263373.33772421 1028454.51823533 1062454.86789143
 1128794.22251689 1202771.54165208 1155550.37874699 1412004.95380795
 1267973.40484512 1273172.16560066 1193345.42617428 1313998.85161936
 1192920.97513652 1189698.3320452  1121988.4301579  1144473.52168047
 1087342.08746111 1135488.7650497  1

학습된 모델을 이용해 평가 데이터셋의 예측값 출력
sklearn을 이용해 선언된 모델은 predict() 메소드를 이용해 간단한 결과값 예측 가능

In [155]:
sample_submission['Weekly_Sales'] = prediction
sample_submission.to_csv('./dataset/submission.csv',index = False)
sample_submission

id  Weekly_Sales
0      1  1.285507e+06
1      2  1.232370e+06
2      3  1.122500e+06
3      4  1.101443e+06
4      5  1.244055e+06
..   ...           ...
175  176  5.852134e+05
176  177  6.729842e+05
177  178  6.736501e+05
178  179  6.092471e+05
179  180  6.100834e+05

[180 rows x 2 columns]